### PR TITLE
Implement GCP notifications processor

### PR DESCRIPTION
**Pull request by ariefrahmansyah**
# TL;DR

Implement notifications processor for GCP.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description

Add new `GcpProcessor` that implements `Processor` interface.

## Tracking Issue
_Remove the '*fixes*' keyword if there will be multiple PRs to fix the linked issue_

fixes https://github.com/flyteorg/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

### CODE REVIEW

File: pkg/async/notifications/factory.go [hunk 1]
```diff
@@ -103,14 +104,31 @@ func NewNotificationsProcessor(config runtimeInterfaces.NotificationsConfig, sco
 			panic(err)
 		}
 		emailer = GetEmailer(config, scope)
+		cloudProvider = common.AWS
```
Comment (pmahindrakar-oss) [inline pkg/async/notifications/factory.go]: Any reason to introduce a new var instead of using config.Type directly

File: pkg/async/notifications/factory.go [hunk 2]
```diff
@@ -103,14 +104,31 @@ func NewNotificationsProcessor(config runtimeInterfaces.NotificationsConfig, sco
...
+	case common.GCP:
+		projectID := config.GCPConfig.ProjectID
+		subscibrer := config.NotificationsProcessorConfig.QueueName
```
Comment (pmahindrakar-oss) [inline pkg/async/notifications/factory.go:109]: nit : subscibrer -> subscriber
Also would be good if we can do some basic validation around config.GCPConfig ,ProjectId , config.NotificationsProcessorConfig and QueuName

File: pkg/async/notifications/implementations/processor.go [hunk 3]
```diff
@@ -57,53 +56,63 @@ func (p *Processor) run() error {
...
-		}
-
-		if err = proto.Unmarshal(notificationBytes, &emailMessage); err != nil {
-			logger.Debugf(context.Background(), "failed to unmarshal to notification object from decoded string[%s] from message [%s] with err: %v", valueString, stringMsg, err)
-			p.systemMetrics.MessageDecodingError.Inc()
-			p.markMessageDone(msg)
-			continue
+		stringMsg := string(msg.Message())
+
+		if p.cloudProvider == common.AWS {
```
Comment (pmahindrakar-oss) [inline pkg/async/notifications/implementations/processor.go]: Can we replace this with switch to be consistent with other places on how the control flow changes for different providers,

Comment (ariefrahmansyah) [reply pkg/async/notifications/implementations/processor.go]: I pushed a new approach which no longer using `cloudProvider` type. Please help take a look.

File: pkg/async/notifications/implementations/gcp_processor.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+	}
+}
+
+func (p *GcpProcessor) StartProcessing() {
+	for {
+		logger.Warningf(context.Background(), "Starting GCP notifications processor")
+		err := p.run()
+		logger.Errorf(context.Background(), "error with running GCP processor err: [%v] ", err)
+		time.Sleep(async.RetryDelay)
+	}
```
Comment (pmahindrakar-oss) [inline pkg/async/notifications/implementations/gcp_processor.go:37]: You can use wait.UntilWithContext from apimachinery

File: pkg/async/notifications/implementations/gcp_processor.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+func (p *GcpProcessor) run() error {
```
Comment (pmahindrakar-oss) [inline pkg/async/notifications/implementations/gcp_processor.go:40]: nit : can we pass in the context and use it in places where we are using context.Background

**Review by pmahindrakar-oss — COMMENTED** [review]
Can you also add some integration test details on how you tested this on GCP and results for reference. 
Thanks

File: pkg/async/notifications/implementations/gcp_processor.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+	// According to https://github.com/NYTimes/gizmo/blob/f2b3deec03175b11cdfb6642245a49722751357f/pubsub/pubsub.go#L36-L39,
+	// the channel backing the subscriber will just close if there is an error. The call to Err() is needed to identify
+	// there was an error in the channel or there are no more messages left (resulting in no errors when calling Err()).
+	if err = p.sub.Err(); err != nil {
+		p.systemMetrics.ChannelClosedError.Inc()
+		logger.Warningf(context.Background(), "The stream for the subscriber channel closed with err: %v", err)
+	}
+
+	// If there are no errors, nil will be returned.
+	return err
```
Comment (katrogan) [inline pkg/async/notifications/implementations/gcp_processor.go:73]: I realized we do this in the aws processor too, but I think this is not great go. can you change it to an explicit `return nil`?

File: pkg/async/notifications/implementations/gcp_processor.go [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+func (p *GcpProcessor) StopProcessing() error {
+	// Note: If the underlying channel is already closed, then Stop() will return an error.
+	err := p.sub.Stop()
+	if err != nil {
+		p.systemMetrics.StopError.Inc()
+		logger.Errorf(context.Background(), "Failed to stop the subscriber channel gracefully with err: %v", err)
+	}
+	return err
```
Comment (katrogan) [inline pkg/async/notifications/implementations/gcp_processor.go:90]:
```suggestion
	return nil
```

File: pkg/async/notifications/implementations/gcp_processor.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+		logger.Errorf(context.Background(), "failed to mark message as Done() in processor with err: %v", err)
+	}
+}
+
+func (p *GcpProcessor) StopProcessing() error {
+	// Note: If the underlying channel is already closed, then Stop() will return an error.
+	err := p.sub.Stop()
+	if err != nil {
+		p.systemMetrics.StopError.Inc()
+		logger.Errorf(context.Background(), "Failed to stop the subscriber channel gracefully with err: %v", err)
```
Comment (katrogan) [inline pkg/async/notifications/implementations/gcp_processor.go:88]:
```suggestion
		logger.Errorf(context.Background(), "Failed to stop the subscriber channel gracefully with err: %v", err)
		return err
```

File: pkg/async/notifications/implementations/processor.go [hunk 9]
```diff
@@ -2,35 +2,19 @@ package implementations
...
 	"context"
-	"time"
-
-	"github.com/flyteorg/flyteadmin/pkg/async"
-
-	"github.com/flyteorg/flyteadmin/pkg/async/notifications/interfaces"
-
 	"encoding/base64"
 	"encoding/json"
+	"time"
```
Comment (katrogan) [inline pkg/async/notifications/implementations/processor.go:7]: mind renaming this file to aws_processor?

File: pkg/async/notifications/factory.go [hunk 10]
```diff
@@ -103,14 +103,30 @@ func NewNotificationsProcessor(config runtimeInterfaces.NotificationsConfig, sco
...
+			sub, err = gizmoGCP.NewSubscriber(context.TODO(), projectID, subscription)
+			if err != nil {
+				logger.Warnf(context.TODO(), "Failed to initialize new gizmo gcp subscriber with config [ProjectID: %s, Subscription: %s] and err: %v", projectID, subscription, err)
+			}
+			return err
+		})
+		if err != nil {
+			panic(err)
+		}
+		emailer = GetEmailer(config, scope)
```
Comment (katrogan) [inline pkg/async/notifications/factory.go:121]: out of curiosity, are you using the sendgrid emailer here? how did you set this up for testing?

File: pkg/async/notifications/implementations/gcp_processor.go [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+func (p *GcpProcessor) run() error {
+	var emailMessage admin.EmailMessage
+	var err error
+
+	for msg := range p.sub.Start() {
+		p.systemMetrics.MessageTotal.Inc()
+
+		if err = proto.Unmarshal(msg.Message(), &emailMessage); err != nil {
```
Comment (katrogan) [inline pkg/async/notifications/implementations/gcp_processor.go:47]: out of curiosity, I know for the AWS implementation we had to do some [massaging](https://github.com/flyteorg/flyteadmin/blob/master/pkg/async/notifications/implementations/processor.go#L64) with the SNS/SQS message format in order to extract the email body from the dequeued subscriber message - is this completely unnecessary with gcp?

**Review by katrogan — COMMENTED** [review]
hi @ariefrahmansyah thanks for this change, it looks great! are you able to test this end to end?

File: pkg/async/notifications/factory.go (hunk 10 above)
Comment (ariefrahmansyah) [reply pkg/async/notifications/factory.go:121]: Yes, we are using SendGrid which can be purchased via GCP Marketplace (https://console.cloud.google.com/marketplace/product/sendgrid-app/sendgrid-email). After registering on SendGrid and verify sender identity we can then use that verified email.

The config is like this:
```notifications:
      emailer:
        body: |
          Execution {{ workflow.project }}/{{ workflow.domain }}/{{ workflow.name }}/{{ name }} has {{ phase }}.
          Details: https://flyte.example.com/console/projects/{{ project }}/domains/{{ domain }}/executions/{{ name }}.
          {{ error }}
        sender: <verified-email>@gojek.com
        subject: 'Flyte: {{ project }}/{{ domain }}/{{ launch_plan.name }} has {{ phase }}'
        emailServerConfig:
          serviceName: sendgrid
          apiKeyEnvVar: SENDGRID_API_KEY```
          
I then manually set SENDGRID_API_KEY as environment variable for flyteadmin container. We may also need to be able to configure flyteadmin environment variable via Helm chart as it's not supported for now: https://github.com/flyteorg/flyte/blob/master/charts/flyte/templates/admin/deployment.yaml#L91-L115

File: pkg/async/notifications/implementations/gcp_processor.go (hunk 11 above)
Comment (ariefrahmansyah) [reply pkg/async/notifications/implementations/gcp_processor.go]: Yes, it's unnecessary in GCP because its formatting is different from AWS. When I try using AWS' parser on GCP Pubsub, I got this error:

```
{"json":{"src":"processor.go:111"},"level":"debug","msg":"failed to unmarshal to notification object message [\u0012\ufffd\u0001\n\u0026\n\u000bflytesnacks\u0012\u000bdevelopment\"\npgb4a56coo\u0018\u0004\"\u000c\u0008\ufffd\ufffd\ufffd\ufffd\u0006\u0010\ufffdԈ\ufffd\u0001*bgs://flyte-test/metadata/propeller/flytesnacks-development-pgb4a56coo/end-node/data/0/outputs.pb] with err: string field contains invalid UTF-8","ts":"2021-09-14T22:17:27Z"}
```

**Comment by ariefrahmansyah** [comment]
> hi @ariefrahmansyah thanks for this change, it looks great! are you able to test this end to end?
Yes, here are some screenshots of notification on my slack and inbox:

image

image

File: pkg/async/notifications/implementations/gcp_processor.go (hunk 11 above)
Comment (katrogan) [reply pkg/async/notifications/implementations/gcp_processor.go]: great, good to know that GCP doesn't require any message parsing

File: pkg/async/notifications/factory.go (hunk 10 above)
Comment (katrogan) [reply pkg/async/notifications/factory.go:121]: thank you for testing! do you mind filing an issue for being able to set SENDGRID_API_KEY in the helm chart? https://github.com/flyteorg/flyte/issues/new/choose

**Review by katrogan — APPROVED** [review]

**Comment by katrogan** [comment]
@pmahindrakar-oss mind taking another look?

**Comment by katrogan** [comment]
hey @ariefrahmansyah mind following the instructions in the failing DCO check to rebase your commits?

**Comment by pmahindrakar-oss** [comment]
Looks good to me too @ariefrahmansyah , can you check the unit test failure.

**Comment by ariefrahmansyah** [comment]
> Looks good to me too. @ariefrahmansyah , can you check the unit test failure.

The unit test is running successfully on my local machine and the code changes shouldn't touch that failing package (`github.com/flyteorg/flyteadmin/pkg/repositories/transformers`). Can you help re-run the jobs? @pmahindrakar-oss @katrogan

**Comment by kumare3** [comment]
@ariefrahmansyah i re-ran it seems like linter issue?

**Comment by ariefrahmansyah** [comment]
> @ariefrahmansyah i re-ran it seems like linter issue?

Thanks. I've fixed the lint issue @kumare3

File: pkg/async/notifications/factory.go (hunk 10 above)
Comment (ariefrahmansyah) [reply pkg/async/notifications/factory.go:121]: Here it is: https://github.com/flyteorg/flyte/issues/1520

**Review by pmahindrakar-oss — APPROVED** [review]
